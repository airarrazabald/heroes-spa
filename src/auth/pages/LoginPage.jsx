import { useContext, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { AuthContext } from '../context/AuthContext';

export const LoginPage = () => {

  const { login } = useContext( AuthContext);

  const navigate = useNavigate();

  const onLogin = ( userName ) => {

      const lastPath = localStorage.getItem('lastPath') || '/';

      login(userName);

      navigate(lastPath, {
          replace: true
      });
  }


  const [inputValue, setInputValue ] =   useState('')

  const onInputChange = ( { target } ) => {
    setInputValue( target.value );
  }

  const onSubmit = (event) => {
    event.preventDefault();

    console.log(inputValue);

    if( inputValue.trim().length <= 5 ) return;

    onLogin(inputValue.trim());

    setInputValue('');
  }

  return (
    <div className="container mt-5 text-center col-xl-4 col-md-6 col-xs-12">
        <h1>Heroes SPA</h1>
        <hr />
        <form onSubmit={ onSubmit  }>
          <input 
            className="form-control"
            type="text"
            placeholder="Ingresa un nombre para identificarte"
            onChange={ onInputChange }
          />
          <button 
            className="btn btn-primary mt-2 form-control"
          >
            Login
          </button>
        </form>
        <div className="alert alert-info mt-2" role="alert">
          El nombre de usuario es solo para visualizar en el navbar de la app
        </div>
    </div>
  )
}
