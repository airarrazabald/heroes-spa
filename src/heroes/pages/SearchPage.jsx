import { useNavigate, useLocation } from 'react-router-dom';
import  queryString from 'query-string';

import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../components';
import { getHeroesByName } from '../helpers';

export const SearchPage = () => {
  const navigate = useNavigate();

  const location = useLocation();

  const { q = '' } = queryString.parse( location.search);

  const heroes = getHeroesByName(q); 

  const showSearch = (q.length === 0);
  const showError = (q.length > 0) && heroes.length === 0;

  const { searchText, onInputChange } = useForm({
    searchText: q,
  });

  const onSearchSubmit = (event) => {
    event.preventDefault();

    navigate(`?q=${ searchText }`);
  }

  return (
    <>
      <h1>Buscador</h1>
      <hr />

      <div className="row">
        <div className="col-xl-5 col-md-4">
          <h4>Buscar</h4>
          <hr />

          <form onSubmit={ onSearchSubmit } aria-label="form">
            <input 
              type="text" 
              placeholder="Buscar un heroe"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={ searchText }
              onChange={ onInputChange }
            />
            <button className="btn btn-outline-primary mt-2">
              Buscar
            </button>
          </form>
        </div>

        <div className="col-xl-7 col-md-8">
          <h4>Resultados</h4>
          <hr />



          <div className="alert alert-primary animate__animated animate__fadeIn"   style={ {display: showSearch ? '' : 'none'} }>Buscar un heroe</div>

          <div className="alert alert-danger animate__animated animate__fadeIn" aria-label="divWarning" style={ {display: showError ? '' : 'none' } }>No existen heroes con { q }</div>

          {
            heroes.map( heroe => <HeroCard key={ heroe.id } {...heroe} />)
          }
        </div>
      </div>
    </>
  )
}
