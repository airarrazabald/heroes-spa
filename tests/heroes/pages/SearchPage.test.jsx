import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { SearchPage } from '../../../src/heroes/pages/SearchPage';

const mockedUseNAvigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNAvigate

} ));

describe('Pruebas en <SearchPage/>', () => { 

    beforeEach(() => jest.clearAllMocks() );

    test('debe de mostrarse correctamente con valores por defecto ', () => {
        const { container } = render(
            <MemoryRouter>
                <SearchPage/>
            </MemoryRouter>
        )

        expect( container ).toMatchSnapshot();
    });


    test('debe de mostrar a Batman y el input con el valor del query string', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage/>
            </MemoryRouter>
        )

        const input = screen.getByRole('textbox');
        expect( input.value  ).toBe('batman');

        const image = screen.getByRole('img');
        expect( image.src  ).toContain('./assets/heroes/dc-batman.jpg');

        const divMessage = screen.getByLabelText('divWarning')
        expect(divMessage.style.display).toBe('none');


    });

    test('debe de mostrar un error si no se encuentra el hero', () => { 

        render(
            <MemoryRouter initialEntries={['/search?q=megaman']}>
                <SearchPage/>
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        expect( input.value  ).toBe('megaman');

        const divMessage = screen.getByLabelText('divWarning')
        expect(divMessage.style.display).toBe('');

    });

    test('debe de llamar el navigate a la pantalla nueva', () => { 

        const inputValue = 'superman';

        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage/>
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        fireEvent.change(input, { target: { name: 'searchText', value: inputValue }} );

        const form = screen.getByLabelText('form');
        fireEvent.submit(form);

        expect( mockedUseNAvigate ).toHaveBeenCalledWith(`?q=${inputValue}`);


    });
})