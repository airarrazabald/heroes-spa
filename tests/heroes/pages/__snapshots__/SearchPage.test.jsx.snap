// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage/> debe de mostrarse correctamente con valores por defecto  1`] = `
<div>
  <h1>
    Buscador
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-xl-5 col-md-4"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Buscar un heroe"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-2"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      class="col-xl-7 col-md-8"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Buscar un heroe
      </div>
      <div
        aria-label="divWarning"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No existen heroes con 
      </div>
    </div>
  </div>
</div>
`;
